internal/pkg/cli/deploy/workload.go

```diff
@@ -367,24 +388,66 @@ func (d *workloadDeployer) uploadContainerImages(out *UploadArtifactsOutput) err
 	if err != nil {
 		return fmt.Errorf("login to image repository: %w", err)
 	}
+
+	var digestsMu sync.Mutex
 	out.ImageDigests = make(map[string]ContainerImageIdentifier, len(buildArgsPerContainer))
+	var labeledBuffers []*syncbuffer.LabeledSyncBuffer
+	g, ctx := errgroup.WithContext(context.Background())
+	cursor := cursor.New()
+	cursor.Hide()
 	for name, buildArgs := range buildArgsPerContainer {
+		// create a copy of loop variables to avoid data race.
+		name := name
+		buildArgs := buildArgs
+
 		buildArgs.URI = uri
 		buildArgsList, err := buildArgs.GenerateDockerBuildArgs(dockerengine.New(exec.NewCmd()))
 		if err != nil {
-			return fmt.Errorf("generate docker build args: %w", err)
-		}
-		// TODO(adi) handle this log in syncBuffer's Print function
-		log.Infof("Building your container image: docker %s\n", strings.Join(buildArgsList, " "))
-		digest, err := d.repository.BuildAndPush(buildArgs)
-		if err != nil {
-			return fmt.Errorf("build and push image: %w", err)
+			return fmt.Errorf("generate docker build args for %q: %w", name, err)
 		}
-		out.ImageDigests[name] = ContainerImageIdentifier{
-			Digest:            digest,
-			CustomTag:         d.image.CustomTag,
-			GitShortCommitTag: d.image.GitShortCommitTag,
+		buf := syncbuffer.New()
+		labeledBuffers = append(labeledBuffers, buf.WithLabel(fmt.Sprintf("Building your container image %q: docker %s", name, strings.Join(buildArgsList, " "))))
+		pr, pw := io.Pipe()
+		g.Go(func() error {
+			defer pw.Close()
+			digest, err := d.repository.BuildAndPush(ctx, buildArgs, pw)
+			if err != nil {
+				return fmt.Errorf("build and push the image %q: %w", name, err)
+			}
+			digestsMu.Lock()
+			out.ImageDigests[name] = ContainerImageIdentifier{
+				Digest:            digest,
+				CustomTag:         d.image.CustomTag,
+				GitShortCommitTag: d.image.GitShortCommitTag,
+			}
+			digestsMu.Unlock()
+			return nil
+		})
+		g.Go(func() error {
+			if err := buf.Copy(pr); err != nil {
+				return fmt.Errorf("copy build and push output for %q: %w", name, err)
+			}
+			return nil
+		})
+	}
+	opts := []syncbuffer.LabeledTermPrinterOption{syncbuffer.WithPadding(paddingInSpacesForBuildAndPush)}
+	if os.Getenv("CI") != "true" {
+		opts = append(opts, syncbuffer.WithNumLines(defaultNumLinesForBuildAndPush))
+	}
+	ltp := d.labeledTermPrinter(os.Stderr, labeledBuffers, opts...)
+	g.Go(func() error {
+		for {
+			if err := ltp.Print(); err != nil {
+				return fmt.Errorf("print logs: %w", err)
+			}
+			if ltp.IsDone() {
+				return nil
+			}
+			time.Sleep(pollIntervalForBuildAndPush)
```

Comment:
```suggestion
    select {
      case <-ctx.Done():
        return nil
      case <-time.After(pollIntervalForBuildAndPush):
        // loop again
   }
```

so this function stops if ctx is cancelled!